File: app/models/ranking.rb
class Ranking
  Row = Data.define(:rank, :user, :count, :buckets)

  def initialize(entries)
    @entries = entries.map(&:to_h)
  end

  def rows
    @rows ||= compute
  end

  def total_count
    rows.sum(&:count)
  end

  def contributor_count
    rows.size
  end

  def empty?
    rows.empty?
  end

  private

  # Competitive rank: ties share their position, the next rank skips (1, 1, 3, ...).
  # Zero-count entries are omitted.
  def compute
    sorted = @entries
      .reject { |e| e[:count].zero? }
      .sort_by { |e| -e[:count] }

    sorted.each_with_index.with_object([]) do |(entry, idx), ranked|
      tied = ranked.any? && entry[:count] == ranked.last.count
      rank = tied ? ranked.last.rank : idx + 1
      ranked << Row.new(rank:, user: entry[:user], count: entry[:count], buckets: entry[:buckets] || [])
    end
  end
end

File: src/ranking.ts
export interface RankingEntry<U = unknown> {
  user: U;
  count: number;
  buckets?: readonly number[];
}

export interface RankingRow<U = unknown> {
  rank: number;
  user: U;
  count: number;
  buckets: readonly number[];
}

export class Ranking<U = unknown> {
  private readonly _rows: readonly RankingRow<U>[];

  constructor(entries: readonly RankingEntry<U>[]) {
    const sorted = entries.filter((e) => e.count !== 0).sort((a, b) => b.count - a.count);

    const ranked: RankingRow<U>[] = [];
    sorted.forEach((entry, idx) => {
      const tied = ranked.length > 0 && entry.count === ranked[ranked.length - 1]!.count;
      const rank = tied ? ranked[ranked.length - 1]!.rank : idx + 1;
      ranked.push({ rank, user: entry.user, count: entry.count, buckets: entry.buckets ?? [] });
    });

    this._rows = ranked;
  }

  get rows(): readonly RankingRow<U>[] {
    return this._rows;
  }

  get totalCount(): number {
    return this._rows.reduce((sum, row) => sum + row.count, 0);
  }

  get contributorCount(): number {
    return this._rows.length;
  }

  get isEmpty(): boolean {
    return this._rows.length === 0;
  }
}
